core/commands/storage/upload/upload/dc_repair_router.go

```diff
@@ -202,44 +212,338 @@ returns the repairer's signed contract to the invoker.`,
+			err := backoff.Retry(func() error {
+				select {
+				case <-rss.Ctx.Done():
+					return nil
+				default:
+					break
+				}
+				host, err := hp.NextValidHost(m.Price)
+				if err != nil {
+					terr := rss.To(sessions.RssToErrorEvent, err)
+					if terr != nil {
+						log.Debugf("original err: %s, transition err: %s", err.Error(), terr.Error())
+					}
+					return nil
+				}
+				guardContract := contractMap[shardHash]
+				guardContract.HostPid = host
+				guardContract.State = guardpb.Contract_RENEWED
+				guardContractBytes, err := proto.Marshal(guardContract)
+				if err != nil {
+					return nil
+				}
+				repairPid := ctxParams.N.Identity
+				hostPid, err := peer.IDB58Decode(host)
+				if err != nil {
+					log.Errorf("shard %s decodes host_pid error: %s", h, err.Error())
+					return err
+				}
+				cb := make(chan error)
+				ShardErrChanMap.Set(guardContract.ContractId, cb)
+				go func() {
+					ctx, _ := context.WithTimeout(rss.Ctx, 10*time.Second)
+					_, err := remote.P2PCall(ctx, rss.CtxParams.N, rss.CtxParams.Api, hostPid, "/storage/upload/init",
+						rss.SsId,
+						rss.Hash,
+						h,
+						m.Price,
+						"",
+						guardContractBytes,
+						-1,
+						m.ShardFileSize,
+						i,
+						repairPid,
+					)
+					if err != nil {
+						cb <- err
+					}
+				}()
+				// host needs to send recv in 30 seconds, or the contract will be invalid.
+				tick := time.Tick(30 * time.Second)
+				select {
+				case err = <-cb:
+					ShardErrChanMap.Remove(guardContract.ContractId)
+					return err
+				case <-tick:
+					return errors.New("host timeout")
+				}
+			}, uh.HandleShardBo)
+			if err != nil {
+				_ = rss.To(sessions.RssToErrorEvent,
+					errors.New("timeout: failed to setup contract in "+uh.HandleShardBo.MaxElapsedTime.String()))
+			}
+		}(shardIndexes[index], shardHash)
+	}
+	updateRepairContract(contracts, contractMap)
+	return statusMeta, nil
+}
+
+func updateRepairContract(contracts []*guardpb.Contract, contractMap map[string]*guardpb.Contract) {
+	for _, contract := range contracts {
+		if contract.State == guardpb.Contract_LOST {
+			if v, ok := contractMap[contract.ShardHash]; ok {
+				contract = v
+				delete(contractMap, contract.ShardHash)
+			}
+		}
+	}
+}
+
+func downloadAndRebuildFile(ctxParams *uh.ContextParams, fileHash string, rs []string) ([]byte, error) {
+	url := fmt.Sprint(strings.Split(ctxParams.Cfg.Addresses.API[0], "/")[2], ":", strings.Split(ctxParams.Cfg.Addresses.API[0], "/")[4])
+	resp, err := shell.NewShell(url).Request("get", fileHash).Option("rs", rs).Send(ctxParams.Ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Close()
+
+	if resp.Error != nil {
+		return nil, resp.Error
+	}
+	return ioutil.ReadAll(resp.Output)
+}
+
+func challengeShards(ctxParams *uh.ContextParams, repairReq *guardpb.RepairContract) error {
+	errChan := make(chan error, len(repairReq.LostShardHash))
+	for _, lostShardHash := range repairReq.LostShardHash {
+		go func() {
+			err := func() error {
```

Comment:
this part of code should be same with existing, can we try to make that part a function, and call it here?

Reply:
Extracted a function of challenge share here.